rustls/src/msgs/handshake.rs

```diff
@@ -1577,35 +1642,142 @@ impl Codec<'_> for ServerEcdhParams {
+        Self {
+            dh_p: PayloadU16::new(params.p.to_vec()),
+            dh_g: PayloadU16::new(params.g.to_vec()),
+            dh_Ys: PayloadU16::new(kx.pub_key().to_vec()),
+        }
+    }
+
+    #[cfg(feature = "tls12")]
+    fn named_group(&self) -> Option<NamedGroup> {
+        fn trim_leading_zeros(buf: &[u8]) -> &[u8] {
+            for start in 0..buf.len() {
+                if buf[start] != 0 {
+                    return &buf[start..];
+                }
+            }
+            &[]
+        }
+
+        FfdheGroup {
+            p: trim_leading_zeros(&self.dh_p.0),
+            g: trim_leading_zeros(&self.dh_g.0),
+        }
+        .named_group()
+    }
 }
 
-impl Codec<'_> for EcdheServerKeyExchange {
+impl Codec<'_> for ServerDhParams {
     fn encode(&self, bytes: &mut Vec<u8>) {
-        self.params.encode(bytes);
-        self.dss.encode(bytes);
+        self.dh_p.encode(bytes);
+        self.dh_g.encode(bytes);
+        self.dh_Ys.encode(bytes);
     }
 
     fn read(r: &mut Reader) -> Result<Self, InvalidMessage> {
-        let params = ServerEcdhParams::read(r)?;
-        let dss = DigitallySignedStruct::read(r)?;
+        Ok(Self {
+            dh_p: PayloadU16::read(r)?,
+            dh_g: PayloadU16::read(r)?,
+            dh_Ys: PayloadU16::read(r)?,
+        })
+    }
+}
+
+#[derive(Debug)]
+pub(crate) enum ServerKeyExchangeParams {
+    Ecdh(ServerEcdhParams),
+    Dh(ServerDhParams),
+}
+
+impl ServerKeyExchangeParams {
+    #[cfg(feature = "tls12")]
+    pub(crate) fn new(kx: &dyn ActiveKeyExchange) -> Self {
+        match kx.group().key_exchange_algorithm() {
+            Some(KeyExchangeAlgorithm::DHE) => Self::Dh(ServerDhParams::new(kx)),
+            Some(KeyExchangeAlgorithm::ECDHE) => Self::Ecdh(ServerEcdhParams::new(kx)),
+            None => panic!("`ActiveKeyExchange` with unknown key exchange algorithm"),
+        }
+    }
+
+    #[cfg(feature = "tls12")]
+    pub(crate) fn pub_key(&self) -> &[u8] {
+        match self {
+            Self::Ecdh(ecdh) => &ecdh.public.0,
+            Self::Dh(dh) => &dh.dh_Ys.0,
+        }
+    }
+
+    pub(crate) fn encode(&self, buf: &mut Vec<u8>) {
+        match self {
+            Self::Ecdh(ecdh) => ecdh.encode(buf),
+            Self::Dh(dh) => dh.encode(buf),
+        }
+    }
 
-        Ok(Self { params, dss })
+    #[cfg(feature = "tls12")]
+    pub(crate) fn named_group(&self) -> Option<NamedGroup> {
+        match self {
+            Self::Ecdh(ecdh) => Some(ecdh.curve_params.named_group),
+            Self::Dh(dh) => dh.named_group(),
+        }
+    }
+}
+
+impl KxDecode<'_> for ServerKeyExchangeParams {
+    fn decode(r: &mut Reader, algo: KeyExchangeAlgorithm) -> Result<Self, InvalidMessage> {
+        use KeyExchangeAlgorithm::*;
+        Ok(match algo {
+            ECDHE => Self::Ecdh(ServerEcdhParams::read(r)?),
+            DHE => Self::Dh(ServerDhParams::read(r)?),
+        })
+    }
+}
+
+#[derive(Debug)]
+pub struct ServerKeyExchange {
```

Comment:
Nit: it would be nice to extract the "boring" changes here (create a `ServerKeyExchange` type and a `ServerKeyExchangeParams` type, renaming the `ServerKeyExchangePayload` variant) into a separate commit, as much as is possible/makes sense.